subsys/power/pm_ctrl.c

```diff
@@ -15,65 +15,35 @@
 #include <logging/log.h>
 LOG_MODULE_DECLARE(power);
 
-struct pm_ctrl_info {
-	int pm_state;
-	atomic_t count;
-};
+static atomic_t power_state_disable_count[SYS_POWER_STATE_MAX];
 
-static struct pm_ctrl_info pm_ctrl[] = {
-	{SYS_PM_LOW_POWER_STATE, ATOMIC_INIT(0)},
-	{SYS_PM_DEEP_SLEEP, ATOMIC_INIT(0)},
-};
-
-void sys_pm_ctrl_disable_state(int state)
+void sys_pm_ctrl_disable_state(enum power_states state)
 {
-	if (state == SYS_PM_LOW_POWER_STATE) {
-		__ASSERT(pm_ctrl[0].count < UINT_MAX,
-				"Low Power state count overflowed\n");
-		atomic_inc(&pm_ctrl[0].count);
-	} else if (state == SYS_PM_DEEP_SLEEP) {
-		__ASSERT(pm_ctrl[1].count < UINT_MAX,
-				"Deep Sleep state count overflowed\n");
-		atomic_inc(&pm_ctrl[1].count);
-	} else {
-		LOG_WRN("\nInvalid PM state");
-	}
+	atomic_val_t v;
+
+	__ASSERT(state < SYS_POWER_STATE_MAX, "Invalid state!\n");
```

Comment:
You are changing a Warning to an ASSERT. What is the reason for that?

Reply:
You cannot enable/disable undefined state. If you try, you will get error.

Reply:
1. Where is that error?

2. If such an error is discovered below, and is solely due to undefined state, then, we can keep the ASSERT here and remove the checks below. This way we can save execution cycles by assuming that the caller supplies a well-defined state...

Reply:
All asserts there are catching API misuse.

The: `__ASSERT(state < SYS_POWER_STATE_MAX, "Invalid power state!");` catches invalid argument.
The `__ASSERT(v > 0, "Power state disable count underflowed!");` and `__ASSERT(v < UINT_MAX, "Power state disable count overflowed!");` caches unpaired enable/disable repeated several times.

Please note, that the assert is an development tool. I is being disabled in the production code.